src/Cassandra/Cluster.cs

```diff
@@ -488,5 +522,77 @@ internal static HostDistance RetrieveDistance(Host host, ILoadBalancingPolicy lb
             host.SetDistance(distance);
             return distance;
         }
+
+        /// <inheritdoc />
+        async Task<PreparedStatement> IInternalCluster.Prepare(
+            IInternalSession session, Serializer serializer, InternalPrepareRequest request)
+        {
+            var lbp = session.Cluster.Configuration.DefaultRequestOptions.LoadBalancingPolicy;
+            var handler = InternalRef.Configuration.PrepareHandlerFactory.Create(serializer);
+            var ps = await handler.Prepare(request, session, lbp.NewQueryPlan(session.Keyspace, null).GetEnumerator()).ConfigureAwait(false);
+            var psAdded = InternalRef.PreparedQueries.GetOrAdd(ps.Id, ps);
+            if (ps != psAdded)
+            {
+                PrepareHandler.Logger.Warning("Re-preparing already prepared query is generally an anti-pattern and will likely " +
+                               "affect performance. Consider preparing the statement only once. Query='{0}'", ps.Cql);
+                ps = psAdded;
+            }
+
+            return ps;
+        }
+        
+        private async Task PrepareAllQueries(Host host)
+        {
+            ICollection<PreparedStatement> preparedQueries = InternalRef.PreparedQueries.Values;
```

Comment:
Now that I think about it, `ConcurrentDictionary<T>.Values` get property is an expensive operation (acquires all locks).

Outside the scope of this ticket, we should use a collection for prepared queries where reading is cheap.

Reply:
created https://datastax-oss.atlassian.net/browse/CSHARP-768